[FLINK-22234][runtime] Read savepoint before creating ExecutionGraph

SteNicholas opened the pull request:
## What is the purpose of the change

*Savepoints are currently read when the `ExecutionGraph` is created. This is causing problems when determining the (max) parallelism the job is run with, because the savepoint may impose additional restrictions. This caused quite a few headaches in FLINK-21844, and further requires the execution vertices to have a mutable max parallelism. `DefaultExecutionGraphFactory` should load the savepoint and then extract from it potential constraints for the `ExecutionGraph` creation.*

## Brief change log

  - *Separate `Checkpoints.loadAndValidateCheckpoint` into `Checkpoints.loadCheckpoint` and `Checkpoints.validateCheckpoint` for loading the savepoint before the `ExecutionGraph` creation.*

## Verifying this change

  - *`CheckpointMetadataLoadingTest` updates `Checkpoints.loadAndValidateCheckpoint` to `Checkpoints.loadCheckpoint` and `Checkpoints.validateCheckpoint`.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

SteNicholas commented:
@flinkbot run azure

tillrohrmann reviewed (commented):
Thanks for creating this PR @SteNicholas. While the PR moves the savepoint creation/reading to before the `ExecutionGraph` creation, it also doubles the time an initial savepoint is read (a second time when the `CheckpointCoordinator` is created). So one idea, which we don't have to do right now but which we should consider is to decouple the `CheckpointCoordinator` from the concrete `ExecutionGraph` instance. That way one could create the `CheckpointCoordinator` before creating an `ExecutionGraph`, reading potential savepoints, checkpoints, extract their meta information and then create an `ExecutionGraph` and register it at the `CheckpointCoordinator` in order to enable checkpointing.

SteNicholas commented:
@tillrohrmann , thanks for your detailed reivew. I have updated the constucture of `CheckpointCoordinator` to replace the `checkpointStorage` parameter with the `checkpointStorageView`, which `checkpointStorageView` is created before the creation of `ExecutionGraph`. This change could only execute to read the savepoint once. Please help to review again for this update.

SteNicholas commented:
@tillrohrmann , @zentol , thanks for your detailed reivew. I have updated the constucture of `CheckpointCoordinator` to replace the `checkpointStorage` parameter with the `checkpointStorageView`, which `checkpointStorageView` is created before the creation of `ExecutionGraph`. This change could only execute to read the savepoint once. Please help to review again for this update addressing your comments.

tillrohrmann commented:
Thanks for updating this PR @SteNicholas. To manage a bit your expectations, I am a bit under water these days. Hence, the review from my side might take a bit of time. Sorry for this.

SteNicholas commented:
@tillrohrmann, thanks for the work of reviewing the pull request.  And if you are available, please help to review again. Thanks.

SteNicholas commented:
@tillrohrmann , @zentol , could you please take a look at this pull request? I have already commented on the issue. Any feedback?

SteNicholas commented:
> Thanks for updating this PR @SteNicholas. To manage a bit your expectations, I am a bit under water these days. Hence, the review from my side might take a bit of time. Sorry for this.

@tillrohrmann , how is your review progress? I have rebased the master branch and please help to review again.
cc @zentol
